crates/typst/src/layout/grid/layout.rs

```diff
@@ -1455,43 +1704,330 @@ impl<'a> GridLayouter<'a> {
+                    pod.backlog = backlog;
+                    cell.measure(engine, self.styles, pod)?.into_frames()
+                }
+            } else {
+                // This row is breakable, so measure the cell normally, with
+                // the initial height and backlog determined previously.
                 let mut pod = self.regions;
-                pod.size.x = self.cell_spanned_width(x, cell.colspan.get());
-
-                let frames = cell.measure(engine, self.styles, pod)?.into_frames();
-
-                // Skip the first region if one cell in it is empty. Then,
-                // remeasure.
-                if let [first, rest @ ..] = frames.as_slice() {
-                    if can_skip
-                        && first.is_empty()
-                        && rest.iter().any(|frame| !frame.is_empty())
-                    {
-                        return Ok(None);
-                    }
+                pod.size = Axes::new(width, height);
+                pod.backlog = backlog;
+                cell.measure(engine, self.styles, pod)?.into_frames()
+            };
+
+            // Skip the first region if one cell in it is empty. Then,
+            // remeasure.
+            if let [first, rest @ ..] = frames.as_slice() {
+                if can_skip
+                    && !unbreakable
+                    && first.is_empty()
+                    && rest.iter().any(|frame| !frame.is_empty())
+                {
+                    return Ok(None);
+                }
+            }
+
+            // Skip frames from previous regions if applicable.
+            let mut sizes = frames
+                .iter()
+                .skip(frames_in_previous_regions)
+                .map(|frame| frame.height())
+                .collect::<Vec<_>>();
+
+            // Don't expand this row more than the cell needs.
+            // To figure out how much height the cell needs, we must first
+            // subtract, from the cell's expected height, the already resolved
+            // heights of its spanned rows. Note that this is the last spanned
+            // auto row, so all previous auto rows were already resolved, as
+            // well as fractional rows in previous regions.
+            // Additionally, we subtract the heights of fixed-size rows which
+            // weren't laid out yet, since those heights won't change in
+            // principle.
+            // Upcoming fractional rows are ignored.
+            // Upcoming gutter rows might be removed, so we need to simulate
+            // them.
+            if rowspan > 1 {
+                // Subtract already covered height from previous rows in the
+                // current region, if applicable.
+                let mut already_covered_height = height_in_this_region;
+
+                // Remove frames which were already covered by previous rows
+                // spanned by this cell.
+                while already_covered_height > Abs::zero()
```

Comment:
Good catch, this is a bit outdated. I used to do this for previous regions as well, but then I switched to just skipping the previous frames, which works much better.

I'll just `continue` if the rowspan is already fully covered by previous rows in the same region, otherwise have a simple subtraction.